src/packages/auth/src/authentication/resolvers/api-key/base-resolver.ts

```diff
@@ -0,0 +1,154 @@
+	gqlEntityType: GraphqlEntityType<G, D>,
+	provider: BackendProvider<D, G>
+) => {
+	const transactional = !!provider.withTransaction;
+
+	@Resolver()
+	abstract class ApiKeyBaseResolver extends createBaseResolver(gqlEntityType, provider) {
+		abstract create(params: CreateOrUpdateHookParams<ApiKeyInputArgs>): Promise<D>;
+		abstract update(params: CreateOrUpdateHookParams<ApiKeyCreateOrUpdateInputArgs>): Promise<D>;
+
+		public async withTransaction<T>(callback: () => Promise<T>) {
+			return provider.withTransaction ? provider.withTransaction<T>(callback) : callback();
+		}
+
+		@Mutation(() => gqlEntityType)
+		async createApiKey(
+			@Arg('data', () => ApiKeyInputArgs) data: ApiKeyInputArgs,
+			@Ctx() context: AuthorizationContext,
+			@Info() info: GraphQLResolveInfo
+		): Promise<G | null> {
+			return this.withTransaction(async () => {
+				const params = {
+					args: { items: [data] },
+					info,
+					context,
+					transactional,
+				};
+
+				const hookParams = await runWritableBeforeHooks<ApiKeyInputArgs>(
+					HookRegister.BEFORE_CREATE,
+					params,
+					'ApiKey'
+				);
+
+				let apiKey;
+				try {
+					apiKey = await this.create(hookParams);
+				} catch (err) {
+					logger.error(err);
+					if (err instanceof ValidationError) throw err;
+					if (err instanceof ForbiddenError)
+						throw new ForbiddenError(
+							'Permission Denied: You do not have permission to create API keys.'
+						);
+
+					throw new AuthenticationError(
+						'Create unsuccessful: You do not have permission to perform this action.'
+					);
+				}
+
+				if (gqlEntityType.fromBackendEntity) {
+					return gqlEntityType.fromBackendEntity(apiKey);
+				}
+
+				return apiKey as any;
+			});
+		}
+
+		@Mutation(() => gqlEntityType)
+		async updateApiKey(
+			@Arg('data', () => ApiKeyCreateOrUpdateInputArgs) data: ApiKeyCreateOrUpdateInputArgs,
+			@Ctx() context: AuthorizationContext,
+			@Info() info: GraphQLResolveInfo
+		): Promise<G | null> {
+			return this.withTransaction(async () => {
+				const params = {
+					args: { items: [data] },
+					info,
+					context,
+					transactional,
+				};
+
+				let apiKey;
+				try {
+					const hookParams = await runWritableBeforeHooks<ApiKeyCreateOrUpdateInputArgs>(
+						HookRegister.BEFORE_UPDATE,
+						params,
+						'ApiKey'
+					);
+					apiKey = await this.update(hookParams);
+				} catch (err) {
+					logger.error(err);
+					if (err instanceof ValidationError) throw err;
+					if (err instanceof ForbiddenError)
+						throw new ForbiddenError(
+							'Permission Denied: You do not have permission to update an API key.'
+						);
+					throw new AuthenticationError(
+						`Update unsuccessful: You do not have permission to perform this action.`
+					);
+				}
+
+				if (gqlEntityType.fromBackendEntity) {
+					return gqlEntityType.fromBackendEntity(apiKey);
+				}
+
+				return apiKey as any;
+			});
+		}
+	}
+
+	return ApiKeyBaseResolver;
```

Comment:
This is the abstract "hey you wanna make your own API key resolver you gotta follow this class"